pkg/workloadmeta/store.go

```diff
@@ -0,0 +1,382 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2016-present Datadog, Inc.
+
+package workloadmeta
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	"github.com/DataDog/datadog-agent/pkg/errors"
+	"github.com/DataDog/datadog-agent/pkg/status/health"
+	"github.com/DataDog/datadog-agent/pkg/util/log"
+	"github.com/DataDog/datadog-agent/pkg/util/retry"
+)
+
+var (
+	globalStore *Store
+	initOnce    sync.Once
+)
+
+const (
+	retryCollectorInterval = 30 * time.Second
+	pullCollectorInterval  = 5 * time.Second
+	eventBundleChTimeout   = 1 * time.Second
+	eventChBufferSize      = 50
+)
+
+type subscriber struct {
+	name   string
+	ch     chan EventBundle
+	filter *Filter
+}
+
+// Store is a central storage of metadata about workloads. A workload is any
+// unit of work being done by a piece of software, like a process, a container,
+// a kubernetes pod, or a task in any cloud provider.
+type Store struct {
+	storeMut sync.RWMutex
+	store    map[Kind]map[string]Entity
+
+	subscribersMut sync.RWMutex
+	subscribers    []subscriber
+
+	candidates map[string]Collector
+	collectors map[string]Collector
+
+	eventCh chan []Event
+}
+
+// NewStore creates a new workload metadata store, building a new instance of
+// each registered collector. Call Run to start the store and its collectors.
+func NewStore() *Store {
+	candidates := make(map[string]Collector)
+	for id, c := range collectorCatalog {
+		candidates[id] = c()
+	}
+
+	return &Store{
+		store:       make(map[Kind]map[string]Entity),
+		subscribers: []subscriber{},
+
+		candidates: candidates,
+		collectors: make(map[string]Collector),
+		eventCh:    make(chan []Event, eventChBufferSize),
+	}
+}
+
+// Run starts the workload metadata store.
+func (s *Store) Run(ctx context.Context) {
+	retryTicker := time.NewTicker(retryCollectorInterval)
+	pullTicker := time.NewTicker(pullCollectorInterval)
+	health := health.RegisterLiveness("workloadmeta-store")
+
+	// Start collectors immediately
+	s.startCandidates(ctx)
+
+	// Start a pull immediately to fill the store without waiting for the
+	// next tick.
+	pullCtx, pullCancel := context.WithTimeout(ctx, pullCollectorInterval)
+	s.pull(pullCtx)
+
+	log.Info("workloadmeta store initialized successfully")
+
+	go func() {
+		for {
+			select {
+			case <-health.C:
+
+			case <-pullTicker.C:
+				// pullCtx will always be expired at this point
+				// if pullTicker has the same duration as
+				// pullCtx, so we cancel just as good practice
+				pullCancel()
+
+				pullCtx, pullCancel = context.WithTimeout(ctx, pullCollectorInterval)
+				s.pull(pullCtx)
+
+			case evs := <-s.eventCh:
+				s.handleEvents(evs)
```

Comment:
I can be good to add this case at the end.

its remember me an investigation where a `for select` loop never end properly with `ctx.Done()` because it received continuously events. It was coming from the fact that in golang the `select` take in order the `case` so if you have always event in the channel `s.eventCh`,  the cases `retryTicker.C` and `ctx.Done()` will never be reach.

Reply:
I don't mind moving it, but it does not sound accurate to me. What the [spec](https://golang.org/ref/spec#Select_statements) says:

> If one or more of the communications can proceed, a single one that can proceed is chosen via a uniform pseudo-random selection